evap/static/js/datagrid.js

```diff
@@ -0,0 +1,284 @@
+        for (const row of this.rows) {
+            const isDisplayedBySearch = searchWords.every(searchWord => {
+                return row.search.some(rowWord => rowWord.includes(searchWord));
+            });
+            const isDisplayedByFilters = Object.entries(this.state.filter).every(([name, filterValues]) => {
+                const matchesAtLeastOneValue = filterValues.some(filterValue => {
+                    return row.filter.get(name).some(rowValue => rowValue === filterValue);
+                });
+                return filterValues.length === 0 || matchesAtLeastOneValue;
+            });
+            row.isDisplayed = isDisplayedBySearch && isDisplayedByFilters;
+        }
+    }
+
+    order(sorting) {
+        this.state.sorting = sorting;
+        this.applyOrder();
+        this.update();
+    }
+
+    applyOrder() {
+        for (const header of this.headers.values()) {
+            header.removeClass("col-sorting-asc col-sorting-desc");
+        }
+        for (const [column, ordering] of this.state.sorting) {
+            this.headers.get(column).addClass(`col-sorting-${ordering}`);
+        }
+
+        this.rows.sort((a, b) => {
+            for (const [column, ordering] of this.state.sorting) {
+                if (a.order.get(column) < b.order.get(column)) {
+                    return ordering === "asc" ? -1 : 1;
+                } else if (a.order.get(column) > b.order.get(column)) {
+                    return ordering === "asc" ? 1 : -1;
+                }
+            }
+            return 0;
+        });
+    }
+
+    update() {
+        this.container.children().remove();
+        const elements = this.rows
+            .filter(row => row.isDisplayed)
+            .map(row => row.element);
+        this.container.append($(elements));
+        this.saveToStorage();
+    }
+
+    loadFromStorage() {
+        this.state = Object.assign({
+            search: "",
+            filter: {},
+            sorting: this.defaultSorting,
+        }, JSON.parse(localStorage.getItem(this.storageKey)));
+        this.syncFilterStateWithInputs();
+        this.applyFilter();
+        this.applyOrder();
+        this.update();
+    }
+
+    saveToStorage() {
+        localStorage.setItem(this.storageKey, JSON.stringify(this.state));
+    }
+
+    syncFilterStateWithInputs() {
+        this.searchInput.val(this.state.search);
+    }
+}
+
+export class TableGrid extends DataGrid {
+    attachTo(table) {
+        table.find("thead").find("th.col-sorting").each((index, header) => {
+            const position = $(header).index();
+            this.headers.set(position, $(header));
+        });
+        super.attachTo(table.find("tbody"));
+    }
+
+    findSearchableCells(row) {
+        return row.children().get();
+    }
+
+    addFilterRowData(row, data) {}
+
+    findCellByColumn(row, column) {
+        return row.children().eq(column);
+    }
+
+    get defaultSorting() {
+        if (this.columns.length > 0) {
```

Comment:
That's in there because grids without sorting (like the user grid) will still be processed like other grids: `applyOrder` will be called. Passing `sorting = []` makes this a method a no-op. But passing a value like `[[undefined, "asc"]]` will throw an error :(

It is more a guard to access `this.columns` within bounds

Reply:
why would we pass `[[undefined, "asc"]]`?

Reply:
....to expand on that, I understand the idea with guarding the access with a bounds check. however, I'd rather make the code throw some error in that case instead of silently eating our programming errors. But that's probably a personal preference?

Reply:
misunderstanding. `columns` are actually only the columns that are used for sorting, so with a better name this makes 100% sense.